File: Traveling_is_Fun.py
#!/bin/python

import math
import os
import random
import re
import sys



#
# Complete the 'connectedCities' function below.
#
# The function is expected to return an INTEGER_ARRAY.
# The function accepts following parameters:
#  1. INTEGER n
#  2. INTEGER g
#  3. INTEGER_ARRAY originCities
#  4. INTEGER_ARRAY destinationCities
#


#Find root of set
def findRoot(sets, i):
    if sets[i] == i:
        return i
    sets[i:i+1] = [findRoot(sets, sets[i])]
    return sets[i]


def connectedCities(n, g, originCities, destinationCities):

    sets = [i for i in range(n+1)]
    ans = [0]*(len(originCities))


    #Union Connected Cities
    for i in range(g+1, n+1):
        mul = i
        # Union i and i's multiples(mul) less than n.
        while mul <= n :
            root_mul = findRoot(sets, mul)
            root_i = findRoot(sets, i)
            if root_mul != root_i:
                sets[root_mul] = root_i
            mul += i

    #Compare origin's root and destination's root.
    for i in range(len(originCities)):
        ans[i] = 1 if (findRoot(sets, originCities[i]) == findRoot(sets, destinationCities[i])) else 0

    return ans



if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(raw_input().strip())

    g = int(raw_input().strip())

    originCities_count = int(raw_input().strip())

    originCities = []

    for _ in xrange(originCities_count):
        originCities_item = int(raw_input().strip())
        originCities.append(originCities_item)

    destinationCities_count = int(raw_input().strip())

    destinationCities = []

    for _ in xrange(destinationCities_count):
        destinationCities_item = int(raw_input().strip())
        destinationCities.append(destinationCities_item)

    result = connectedCities(n, g, originCities, destinationCities)

    fptr.write('\n'.join(map(str, result)))
    fptr.write('\n')

    fptr.close()

File: TravelingIsFun.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;



class Result {

    /*
     * Complete the 'connectedCities' function below.
     *
     * The function is expected to return an INTEGER_ARRAY.
     * The function accepts following parameters:
     *  1. INTEGER n
     *  2. INTEGER g
     *  3. INTEGER_ARRAY originCities
     *  4. INTEGER_ARRAY destinationCities
     */
    public static int findRoot(List<Integer> sets, int i){
        if(sets.get(i) == i)
            return i;
        sets.set(i,findRoot(sets,sets.get(i)));
        return sets.get(i);
    }

    public static List<Integer> connectedCities(int n, int g, List<Integer> originCities, List<Integer> destinationCities) {
        List<Integer> sets = new ArrayList<Integer>();
        List<Integer> ans = new ArrayList<Integer>();

        for(int i=0; i<n+1 ;i++)
            sets.add(i);
        for(int i=0; i<originCities.size();i++)
            ans.add(0);

        /* Union connected cities*/
        for(int i=g+1 ; i<n+1 ; i++){
            int mul = i;
            /* Union i and i's multiples(mul) less than n */
            while (mul <= n) {
                int root_mul = findRoot(sets, mul);
                int root_i = findRoot(sets, i);
                if (root_mul != root_i)
                    sets.set(root_mul, root_i);
                mul += i;
            }
        }

        /* Compare origin's root and destination's root */
        for(int i=0; i<originCities.size(); i++)
            ans.set(i, findRoot(sets,originCities.get(i))==findRoot(sets,destinationCities.get(i)) ? 1 : 0);

        return ans;


    }

}

public class TravelingIsFun {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = Integer.parseInt(bufferedReader.readLine().trim());

        int g = Integer.parseInt(bufferedReader.readLine().trim());

        int originCitiesCount = Integer.parseInt(bufferedReader.readLine().trim());

        List<Integer> originCities = IntStream.range(0, originCitiesCount).mapToObj(i -> {
            try {
                return bufferedReader.readLine().replaceAll("\\s+$", "");
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        })
                .map(String::trim)
                .map(Integer::parseInt)
                .collect(toList());

        int destinationCitiesCount = Integer.parseInt(bufferedReader.readLine().trim());

        List<Integer> destinationCities = IntStream.range(0, destinationCitiesCount).mapToObj(i -> {
            try {
                return bufferedReader.readLine().replaceAll("\\s+$", "");
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        })
                .map(String::trim)
                .map(Integer::parseInt)
                .collect(toList());

        List<Integer> result = Result.connectedCities(n, g, originCities, destinationCities);

        bufferedWriter.write(
                result.stream()
                        .map(Object::toString)
                        .collect(joining("\n"))
                        + "\n"
        );

        bufferedReader.close();
        bufferedWriter.close();
    }
}
